### PR TITLE
Add skin temperature limit for Renoir and some small improvements

### LINKED ISSUE #82: Add CPU Skin Temperature Limit for Renoir
Hello, 

I would like to set CPU Skin Temperature Limit.

### Why do we need it?
PPT LIMIT APU (STAPM Limit Offset 24) gets applied if SST VALUE APU (STAPM Limit Offset 92) does exceed SST LIMIT APU (STAPM Limit Offset 88, 0x38 on Renoir)
Skin temperature name in HWiNFO is "CPU Skin Temperature" and gets set to 45° on my notebook. I need 50°C for long workloads to prevent 15W PPT LIMIT APU

### What can be done?

To fix this we need to change SST LIMIT APU. For higher accepted temperatures we could use higher PPT LIMIT APU, but I did find a way to change this Power limit, which is not important at all, it would be enough to add SST LIMIT APU.

To Change Skin Temperature Limit, on Renoir ec address 0x38 will accept an UInt. But the value in °C need to be multiplied by 256 if you write in the same 0x38 address. Looks like write address is maybe 8 bits off 0x38.

### Should I create a pull request?

I didn't prepare a Developer Environment for this project, so I could create a branch without building it. Would you like to have an untested branch?

Kind Regards
Falco

### PR DESCRIPTION
**Pull request by Falcosc**
Resolves #82

### CODE REVIEW

File: lib/api.c [hunk 1]
```diff
@@ -369,3 +369,14 @@ EXP int CALL set_prochot_deassertion_ramp(ryzen_access ry, uint32_t value) {
 		break;
 	}
 }
+
+EXP int CALL set_skin_temp(ryzen_access ry, uint32_t value){
+	value *= 256;
```
Comment (Falcosc) [inline lib/api.c:374]: @FlyGoat I don't know why Renoir does need this multiplier and not sure if Picasso does have a skin temp setting at all.

But for Renoir it does work great.

**Comment by FlyGoat** [comment]
I'll do some tests when possible.

If all looks good I'll make a release together with Cezanne support.

There is a ABI change so I'll bump the major version.

**Comment by Falcosc** [comment]
let me update it first, because I just found PPT LIMIT APU
And I found wrong address for SetPSI0SocCurrentLimit and wrong doc for SetProchotDeassertionRampTime

**Comment by Falcosc** [comment]
skin-temp, skin-limit and psi0soc-current do now work great on Renoir.

Still don't know how prochot-deassertion-ramp does work, at least it is nothing in s or ms. So I changed the documentation.
If I use values below ~256 nothing happens: feels like limit after prochot for values lower then 256 does exceed my thermal capacity right after an prochot. Which means after a fixed amount of prochot the limit gets so loose that power spikes right from 4W (prochot power limit) up to max power draw possible for my current tctl temperature (maybe 25W).
But larger values does limit my power draw after prochot:
- before prochot 45W to 30W depending on temperature until prochot kicks in
- 4W during prochot
- after prochot gets released I have 15W  for value ~300 or 5W for values like ~20.000
- after prochot gets released limit stays for at least 10 minutes (didn't test more then 10 minutes for the small value of 300)

After 10 minutes running cinebench at 4W I did set prochot values like 64 which is so low that right after applying my power draw goes up to 45W

I did then a run with 20.000 which does cool down my laptop very fast because for so large value I get only ~5W after prochot gets release. Based on the fast falling temperature I expect an change in power draw. But after prochot limit stays the same at ~5W until I lower the prochot-deassertion-ramp value

**Comment by Falcosc** [comment]
By the way, I was wrong with my assumption about the relation between PPT APU and skin temperature. 
Yes skin temperature did throttle to 15W, 
yes I found a way to increase my 15W PPT APU limit to any other value with --apu-limit 
and yes the new value was reported in the power table. 
But as it turns out, my skin temperature power limit is still 15W.

So in the end I did a rebase to rename --skin-limit to --apu-limit

Test is successful.

**Comment by FlyGoat** [comment]
I had fixed names of options and sepreate changes into multiple commit on current master branch, chould you please review it?

**Comment by Falcosc** [comment]
I did review it by rebasing my commits. You did only forget our readme.

**Comment by Falcosc** [comment]
I did move return code handling and the 0 value support into a new pull request, so only doc changes are open for this PR

**Comment by Falcosc** [comment]
Found the skin temperature dGPU on 0x39
But like --apu-limit it does only change the ptable entry and does not have an effect on my device (usage value for APU limit and dGPU temperature are both reportet as 0)

**Comment by FlyGoat** [comment]
Thanks.LGTM.
Btw A+A sharing means the system AMD used to allocate power plan for APU and dGPU.

As per my reverse engineering outcome this option changes PPT slow for APU part.

**Comment by Falcosc** [comment]
> 
> 
> Thanks.LGTM.
> Btw A+A sharing means the system AMD used to allocate power plan for APU and dGPU.

Is it a common therm? I never heard it, doesn't find it online and worry that we will cause a lot of questions by the users and integrators of this argument.

> 
> As per my reverse engineering outcome this option changes PPT slow for APU part.

Do you have reverse engineering results of Renoir as well? I thought you just know about Picasso.

- Could you share some results, please? I am still searching for some GPU related things.
- If you found the PPT slow for APU strings in Picasso we may, could figure out if this feature is existing there to

**Comment by FlyGoat** [comment]
> > Thanks.LGTM.
> > Btw A+A sharing means the system AMD used to allocate power plan for APU and dGPU.
> 
> Is it a common therm? I never heard it, doesn't find it online and worry that we will cause a lot of questions by the users and integrators of this argument.

The term is used in AGESA CBS option.

> 
> > As per my reverse engineering outcome this option changes PPT slow for APU part.
> 
> Do you have reverse engineering results of Renoir as well? I thought you just know about Picasso.

Will share tomorrow.


> 
>     * Could you share some results, please? I am still searching for some GPU related things.
> 
>     * if you have string from the reverse engineering, I would like to have them in the argument description because I expect questions and I had trouble mapping the parameter to PM Table entries during test of existing parameters.
> 
>     * If you found the PPT slow for APU strings in Picasso, maybe we could figure out if this feature is existing there to

**Pull request merged**